include/json/allocator.h

```diff
@@ -38,7 +38,7 @@ template <typename T> class SecureAllocator {
    * The pointer argument is tagged as "volatile" to prevent the
    * compiler optimizing out this critical step.
    */
-  void deallocate(volatile pointer p, size_type n) {
+  void deallocate(pointer p, size_type n) {
```

Comment:
```cpp
$ c++filt __Z10deallocatePvj
deallocate(void*, unsigned int)

$ c++filt __Z10deallocatePVvj
deallocate(void volatile*, unsigned int)
```
This seems to be a change in public API, so it would require a bump in the soversion which we try to avoid.

How important is it to drop `volatile`?

Reply:
So far I have just looked at compiler output to find out what has been deprecated. Now I had a look at what the standard actually says.
http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2018/p1152r0.html

There is something interesting under the proposals section [proposals](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2018/p1152r0.html#prop)
```
Deprecate (and eventually remove) non-reference and non-pointer volatile parameters.
Deprecate (and eventually remove) const as well as volatile return values.
References and pointers to volatile data remain valid.
```
But as "p" is a pointer this shouldn't be a problem. And if I replace the pointer type with its definition "T*", the warning disappears.
So this looks like it might be a problem with gcc10 not expanding the type before it does its warning

Reply:
Clang has the same behavior.
Here is a minimal example of the warnings
https://godbolt.org/z/Tr1Ka6

Replacing 'volatile pointer p' with 'volatile T* p' clears the warning

Reply:
> Replacing 'volatile pointer p' with 'volatile T* p' clears the warning

That's because they're different types. You have changed "volatile pointer to T" into "pointer to volatile T". The warning returns if you change the godbolt demo to `T* volatile p`.

It's not about the typedef.

Reply:
Wow. That's annoying.

Reply:
Oh wow, you are [right](https://godbolt.org/z/qPzrW4). There are some rules here about how volatile applies, and how using expands that isn't intuitive to me.
What you want is to have the memory be volatile right? Because that is what is getting zeroed.
So it should be "volatile T*"